### PR TITLE
Fix MMR Proof

### PR DESCRIPTION
**Pull request by clearloop**
## Steps

+ [x] Simplify the data structure in the MMR database
+ [x] Port MMR proof to FFI from rust
+ [x] Add proof to HTTP api
+ [x] Update the MMR proof usage of Golang
+ [x] fix `mmr_size` [#177][#177]
+ [x] every MMR match the pair block number
+ [x] `gen_proof` use member pos, limit by last leaf index
+ [x] `mmr_size` of `gen_proof` should generate from the last leaf
+ [x] Generate mmr list
+ [x] More logs
+ [x] Scale codec API

[#177]: https://github.com/darwinia-network/darwinia-common/issues/177

### CODE REVIEW

File: mmr/ffi.rs [hunk 1]
```diff
@@ -25,22 +25,18 @@ pub extern "C" fn run() -> i32 {
...
-                    .iter()
-                    .map(|item| H256::hex(item))
-                    .collect::<Vec<String>>()
-                    .join(",")
-                    .as_bytes(),
-            )
-            .unwrap();
-        }
-        return CString::new("").unwrap();
+    let mmr = MMR::<_, MergeHash, _>::new(Runner::mmr_size(*leaves.iter().max().unwrap()), store);
```
Comment (HackFisher) [inline mmr/ffi.rs:28]: Need to check whether ```leaves.iter().max()``` is ```block count```

Comment (clearloop) [reply mmr/ffi.rs]: Will add a new argument `block_count` in this function, the relayer client(`dj`) will store it when the proposal starts, and pass to here every round

File: mmr/runner.rs [hunk 2]
```diff
@@ -72,37 +58,40 @@ impl Runner {
...
-        // gen mmr proof
-        let proof = mmr
-            .gen_proof(vec![rpos])?
-            .proof_items()
-            .iter()
-            .map(|h| h.hex())
-            .collect::<Vec<String>>()
-            .join(",");
+        let mut mmr =
+            MMR::<_, MergeHash, _>::new(cmmr::leaf_index_to_mmr_size(pnumber as u64) - 1, store);
```
Comment (clearloop) [inline mmr/runner.rs:86]: This part might have ambiguities.

If use the `mmr_size` to generate MMR before we push the `new leaf` into the store, an Error `MMR(InconsistentStore) ` will be fired, So if we shoud:

+ use `leaf_index_to_mmr_size(pnumber as u64 - 1)` when we push a new leaf into store
+ use `leaf_index_to_mmr_size(pnumber as u64)` while generating proof

Comment (clearloop) [reply mmr/runner.rs:86]: Shoud not be `mmr_size - 1`, shoud be `last_mmr_size + 1`

https://github.com/nervosnetwork/merkle-mountain-range/blob/ea232474496bdf072d7de33297fd54f1ef8df599/src/mmr.rs#L70

Comment (HackFisher) [reply mmr/runner.rs:86]: should be last_mmr_size, self.mmr_count()

Comment (HackFisher) [reply mmr/runner.rs:86]: https://github.com/darwinia-network/shadow/issues/34

Comment (HackFisher) [reply mmr/runner.rs:86]: leaf_index_to_mmr_size is used when gen_proof, to retrieve a sub MMR for generating proof for (member_index, last_index), here before append, should use node counts in the store.

Comment (clearloop) [reply mmr/runner.rs:86]: > #34 

#34 is one of the BUGs, and there must be more fishy BUGs...The common tests work fine(using the database this repo implemented), but the production test doesn't.

File: mmr/runner.rs [hunk 3]
```diff
@@ -72,37 +59,38 @@ impl Runner {
...
-
-        // gen mmr proof
-        let proof = mmr
-            .gen_proof(vec![rpos])?
-            .proof_items()
-            .iter()
-            .map(|h| h.hex())
-            .collect::<Vec<String>>()
-            .join(",");
+        let mut mmr = MMR::<_, MergeHash, _>::new(self.mmr_count().unwrap_or(0) as u64, store);
```
Comment (clearloop) [inline mmr/runner.rs:85]: The result of `self.mmr_count()` is equal to `last_mmr_size` while `last_leaf_index > 0`, select from database directly will be more efficient.

**Comment by clearloop** [comment]
The scale codec API should be placed in the next PR, this PR should be merged after testing the proof is correct with @yanganto [#167][#167]

## Proofs

> Ref to the guide of [#170][#170]
```
	             | Global:             | Current:
	             |              R3     |               R3
	             |            /   \    |             /   \
	             |           c     f   |            c     f
	             |          / \   / \  |           / \   / \
	             |         a   b d   e |          a   b d   e
	Block Number |         0   1 2   3 |          0   1 2   3
```

| Round | Member | Leaf | Proof |
|-|-|-|-|
| 0 | 0 | 3 | `["88e96d4537bea4d9c05d12549907b32561d3bf31f45aae734cdc119f13406cb6","25c0795608c645708681e19579e2f99666725f7c70bd3863efcfe7dfef25d6fd"]` |
|1|2|3|`["23adf5a3be0f5235b36941bcb29b62504278ec5b9cdfa277b992ba4a7a3cd3a2","58991bd1987e8b65ddcaf584128e2833ff5fc7209aa9de585fc5138e7ce708cc"]`|
|2|0|2|`["88e96d4537bea4d9c05d12549907b32561d3bf31f45aae734cdc119f13406cb6","3d6122660cc824376f11ee842f83addc3525e2dd6756b9bcf0affa6aa88cf741"]`|


## How to get proof yourselves

```
# Run shadow with rust logs
$ RUST_LOG=mmr shadow run --fetch

# Post the proposal API
$ curl -H "content-type: application/json" --request POST --data '{ "members": [0], "last_leaf": 3 }' http://localhost:3001/api/v1/proposal

# The swagger docs
http://localhost:3001/swagger/index.html
```

[#167]: https://github.com/darwinia-network/darwinia-common/pull/167
[#170]: https://github.com/darwinia-network/darwinia-common/pull/170

**Comment by HackFisher** [comment]
@clearloop encountered ```Failed to connect to localhost port 3001: Connection refused```

```
promote:target denny$ RUST_LOG=mmr ./shadow run --fetch
Please input your infura key: XXX
Shadow RPC service start at 3000
Shadow HTTP service start at 3001
[GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

[GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
 - using env:	export GIN_MODE=release
 - using code:	gin.SetMode(gin.ReleaseMode)

Please input your infura key: [2020-07-10T02:57:21Z INFO  mmr::ffi] starting mmr service...
[2020-07-10T02:57:21Z WARN  mmr::runner] Could not find block: 0 in cache
[2020-07-10T02:57:21Z TRACE mmr::runner] MMR service restarting after 10s...
[2020-07-10T02:57:31Z TRACE mmr::runner] push eth block number 0 into db succeed.
[2020-07-10T02:57:31Z TRACE mmr::runner] push eth block number 1 into db succeed.
[2020-07-10T02:57:31Z TRACE mmr::runner] push eth block number 2 into db succeed.
[2020-07-10T02:57:31Z TRACE mmr::runner] push eth block number 3 into db succeed.
[2020-07-10T02:57:31Z TRACE mmr::runner] push eth block number 4 into db succeed.
```


```
promote:~ denny$ curl -H "content-type: application/json" --request POST --data '{ "members": [0], "last_leaf": 3 }' http://localhost:3001/api/v1/proposal
curl: (7) Failed to connect to localhost port 3001: Connection refused
```

**Comment by clearloop** [comment]
> the block_number_request could be different with member_index

And the `mmr_proof` fields of blocks `block_number` contains only will be empty(`""`)?

File: internal/core/scale.go [hunk 4]
```diff
@@ -34,16 +34,36 @@ func encodeProof(dnmp eth.DoubleNodeWithMerkleProof) string {
...
+
 	hex := "0x"
 	hex += header.ParentHash[2:]
 	hex += encodeUint(header.TimeStamp, 64)
 	hex += encodeUint(header.Number, 64)
 	hex += strings.ToLower(header.Author[2:])
 	hex += header.TransactionsRoot[2:]
 	hex += header.UnclesHash[2:]
-	hex += "7c"
+	hex += len
```
Comment (clearloop) [inline internal/core/scale.go:66]: Here is the BUG in the last version of scale codec, used a fixed vector length.

Comment (clearloop) [reply internal/core/scale.go:66]:
```
# Run shadow with rust logs
$ RUST_LOG=mmr shadow run --fetch

# Post the proposal API
$ curl -H "content-type: application/json" --request POST --data '{ "members": [0], "last_leaf": 3, "format": "codec" }' http://localhost:3001/api/v1/proposal

# The swagger docs
http://localhost:3001/swagger/index.html
```

File: internal/core/scale.go [hunk 5]
```diff
@@ -26,7 +43,7 @@ func encodeProof(dnmp eth.DoubleNodeWithMerkleProof) string {
 	}
 
 	// pad the length
-	hex += "64"
+	hex += lenToHex(len(dnmp.Proof))
```
Comment (clearloop) [inline internal/core/scale.go:46]: Dynamic proof length

Comment (HackFisher) [reply internal/core/scale.go:46]: This file name looks quite general, but the implementation looks quite domain specific. Are there scale codec go libraries for use, then we may do not need to try to implement here.

File: internal/core/scale.go (hunk 4 above)
Comment (HackFisher) [reply internal/core/scale.go:66]: > Here is the BUG in the last version of scale codec, used a fixed vector length.

For "last version of scale codec", you mean some 3rd party libraries?

Comment (clearloop) [reply internal/core/scale.go:66]: Last time write the scale codec for `off-chain worker`, but not in production, this `7c` -> `len` means the length of `extra_data` field in eth header, the hard code `7c` is only the `extra_data` of block number `8997676`.

File: internal/core/scale.go (hunk 5 above)
Comment (clearloop) [reply internal/core/scale.go:46]: [gossamer][0] supports `scale-codec`, but hard to use, Subscan also supports `scale-codec`, but only for decoding.

[0]: https://github.com/ChainSafe/gossamer

**Pull request merged**